pipeline/common/memory.py

```diff
@@ -0,0 +1,52 @@
+import gc
+import logging
+import os
+from typing import Optional
+
+import psutil
+
+from pipeline.common import format_bytes
+from pipeline.common.logging import get_logger
+
+_memory_logger: Optional[logging.Logger] = None
+_memory_process: Optional[psutil.Process] = None
+_memory_last_bytes: Optional[int] = None
+
+
+def get_memory_string() -> str:
+    """
+    Get the current string representation of the memory usage.
+    """
+    global _memory_process
+    global _memory_last_bytes
+
+    # Lazily initial everything.
+    if not _memory_process:
+        _memory_process = psutil.Process(os.getpid())
+
+    gc.collect()
```

Comment:
Generally triggering garbage collection manually is not a good idea (it interferes with the default efficient behavior of the interpreter and it's not free, so the program will likely become slower).  `log_memory` is a helper function and it has a severe side effect because of this. It is not expected by the user and will slow things down in the client code. I suggest either removing it and calling it manually in places you need it or adding a flag with disabled collection by default, like`gc_collect=False`. If you feel like it's safe to do it from the place in code where you call `log_memory`, pass `gc_collect=True`.